modules/xray/shared/src/main/scala/natchez/xray/XRaySpan.scala

```diff
@@ -0,0 +1,272 @@
+      .split(';')
+      .toList
+      .flatMap(_.split('=') match {
+        case Array(k, v) => List((k, v))
+        case _           => List.empty
+      })
+      .toMap
+
+    foo
+      .get("Root")
+      .map(traceId =>
+        XRayHeader(traceId, foo.get("Parent"), foo.get("Sampled").contains("1"))
+      )
+  }
+
+  private def randomHexString[F[_] : Functor : Random](bytes: Int): F[String] =
+    Random[F].nextBytes(bytes)
+      .map(x => BigInt(1, x).toString(16).reverse.padTo(bytes * 2, '0').reverse)
+
+  private def segmentId[F[_] : Functor : Random]: F[String] =
+    randomHexString[F](8)
+
+  private def traceId[F[_] : Applicative : Clock : Random : Parallel]: F[String] =
+    (Clock[F].realTime, randomHexString[F](12)).parMapN { (t, r) =>
+      s"1-${t.toSeconds.toHexString}-$r"
+    }
+
+  def fromHeader[F[_] : Concurrent : Clock : Random : Parallel](name: String,
+                                                                header: XRayHeader,
+                                                                entry: XRayEntryPoint[F]
+                                                               ): F[XRaySpan[F]] =
+    (segmentId[F], Clock[F].realTime, Ref[F].of(Map.empty[String, Json]), Ref[F].of(List.empty[JsonObject]))
+      .parMapN { (sId, t, fields, children) =>
+        XRaySpan(
+          entry = entry,
+          name = name,
+          segmentId = sId,
+          xrayTraceId = header.traceId,
+          startTime = t,
+          fields = fields,
+          children = children,
+          parent = header.parentId.map(_.asLeft),
+          sampled = header.sampled
+        )
+      }
+
+  def fromKernel[F[_] : Concurrent : Clock : Random : Parallel](name: String,
+                                                                kernel: Kernel,
+                                                                entry: XRayEntryPoint[F]
+                                                               ): F[XRaySpan[F]] =
+    kernel.toHeaders
+      .get(Header)
+      .flatMap(parseHeader)
+      .map(x => fromHeader(name, x, entry))
+      .get
+
+  def fromKernelOrElseRoot[F[_] : Concurrent : Clock : Random : Parallel](name: String,
+                                                                          kernel: Kernel,
+                                                                          entry: XRayEntryPoint[F]
+                                                                         ): F[XRaySpan[F]] =
+    kernel.toHeaders
+      .get(Header)
+      .flatMap(parseHeader)
+      .map(x => fromHeader(name, x, entry))
+      .getOrElse(root(name, entry))
+
+  def root[F[_] : Concurrent : Clock : Random : Parallel](name: String,
+                                                          entry: XRayEntryPoint[F]): F[XRaySpan[F]] =
+    (segmentId[F], traceId[F], Clock[F].realTime, Ref[F].of(Map.empty[String, Json]), Ref[F].of(List.empty[JsonObject]))
+      .parMapN { (sId, tId, t, fields, children) =>
```

Comment:
Out of curiosity any specific reason for the use of `Parallel` here and everywhere?

Reply:
Maybe @bpholt can answer this.

Reply:
`Parallel` isn't strictly necessary; I just wanted to use `parMapN` instead of `mapN`.

Reply:
Right, why is that? :) At a first glance I would think the overhead of `parMapN` exceeds any performance gain.

Reply:
That's definitely possible. I just typically default to `Parallel`. If it's not worth the overhead (or extra complexity) then I'm fine with removing it.

Reply:
Yeah according to https://github.com/typelevel/cats-effect/pull/2239#issue-971957270 there's about any 85% overhead for using `Parallel`. If the ops we were parallelizing were async or CPU-bound it could make sense but I'm pretty sure these are al trivial. I would drop it :)